common: Add FlatpakTransactionLog

smspillaz opened the pull request:
This is an early implementation of transaction logging functionality discussed in #1026 . Currently it logs deployments of installations and uninstallations, which is enough to cover our own usecase.

Fixes #1026

smspillaz commented on common/flatpak-transaction-log.c:
```diff
@@ -0,0 +1,305 @@
+		                                     GCancellable          *cancellable,
+		                                     GError                **error)
+{
+  g_autoptr(JsonNode) node = load_existing_log_if_exists (log->path, cancellable, error);
```
Right now we load the entire log file, append one entry and then write the entire thing out every time we want to log an event. Obviously this incurs quite a lot of overhead.

The benefit of doing it this way is that you minimize the potential for inconsistency, if, eg, one process creates a `FlatpakDir` and holds on to it and meanwhile the underlying transaction log changes. The downside is the overhead.

smspillaz commented on common/flatpak-transaction-log.c:
```diff
@@ -0,0 +1,305 @@
+}
+
+gboolean
+flatpak_transaction_log_commit_deploy_install_event (FlatpakTransactionLog *log,
```
We only log deployments of installations as opposed to updates or appstream updates. Not sure if we want to log those under the some event type or log them separately (or not at all).

alexlarsson replied:
You're the primary consumer of this, but i think its probably useful to record every deploy, rather than just the initial install.

alexlarsson replied:
Maybe with a "previous deployed version" property, which is not there if this is the initial install.

smspillaz commented on common/flatpak-transaction-log.c:
```diff
@@ -0,0 +1,305 @@
+  JsonArray *array = json_node_get_array (node);
+  g_autoptr(JsonNode) event = deploy_install_event (ref, origin);
+
+  json_array_add_object_element (array, json_node_dup_object (event));
```
The log file right now is effectively unbounded. We could limit it to N entries, but the tradeoff is that the log becomes lossy and not as useful to consumers (and especially in our usecase where we want to know if a flatpak was ever uninstalled at some point, that information will be lost at some point, so we end up auto-installing a flatpak that was explicitly uninstalled before).

alexlarsson commented:
I don't like this being a json format, because that means we have to parse it each time, which will slow down over time as it grows. Additionally, updating such a format is racy. 

I would much prefer an append-only (open O_APPEND) format for the log. Such a format is guaranteed atomic by the kernel (if we submit the new changes as one buffer), and it doesn't need parsing over time.

Maybe we could make it an comma separated list of json objects. That way one could wrap it in `[` and ` {} ]` to get a valid json format (the extra `{}` is to match the last comma).

smspillaz commented:
> I would much prefer an append-only (open O_APPEND) format for the log. Such a format is guaranteed atomic by the kernel (if we submit the new changes as one buffer), and it doesn't need parsing over time.

I'm fine doing that - also means we can keep a lot of the work we've done so far here.

> Maybe we could make it an comma separated list of json objects. That way one could wrap it in [ and {} ] to get a valid json format (the extra {} is to match the last comma).

Or `[ ... , null ]`

smspillaz commented:
@alexlarsson I suppose the only thing about doing it this way is that the log file is no longer "standard json", which means that we should not expect other applications to parse it. Should we provide an API in flatpak to read the JSON file into something sensible?

alexlarsson commented:
I dunno, if you're parsing the transaction log you're pretty lowlevel already. I'm not sure how useful a public API for that is...

pwithnall commented:
Some API can always be added for accessing the log later, if people request it. For now, I don’t think that’s worth adding public API for.

pwithnall commented on common/flatpak-dir.c:
```diff
@@ -5363,6 +5364,26 @@ flatpak_dir_deploy (FlatpakDir          *self,
+		          const gchar *ref,
+		          const gchar *origin,
+		          GCancellable *cancellable,
+		          GError      **error)
```
Nitpick: Function arguments are not aligned properly.

smspillaz replied:
Ugh, auto-tabs strike again :)

pwithnall commented on common/flatpak-dir.c:
```diff
@@ -5425,6 +5446,9 @@ flatpak_dir_deploy_install (FlatpakDir   *self,
     goto out;
 
+  if (!log_deploy_install_event (self, ref, origin, cancellable, &local_error))
+    g_warning ("Failed to log deploy-install event: %s", local_error->message);
```
If you’re deliberately not propagating the error here and are returning a warning instead, you should `g_clear_error (&local_error)`.

smspillaz replied:
`local_error` will already be cleared by the use of `g_autoptr` above. Or is the proper convention here to always explicitly clear the error?

smspillaz replied:
(Ah, I see by the comment below that is is the convention)

pwithnall replied:
It’s convention to clear the error here because it’s served its purpose; what we want to avoid is passing the same `local_error` to another call before clearing it, since that would result in setting a `GError` on top of an already-set one.

pwithnall commented on common/flatpak-dir.c:
```diff
@@ -6466,6 +6490,18 @@ flatpak_dir_update (FlatpakDir          *self,
+log_uninstall_event (FlatpakDir  *self,
+                     const gchar *ref,
+                     GCancellable *cancellable,
+                     GError      **error)
```
Nitpick: Function arguments are not aligned properly.

pwithnall commented on common/flatpak-dir.c:
```diff
@@ -6569,6 +6606,9 @@ flatpak_dir_uninstall (FlatpakDir          *self,
     }
 
+  if (!log_uninstall_event (self, ref, cancellable, &logging_error))
+    g_warning ("Failed to log uninstallation: %s", logging_error->message);
```
Missing `g_clear_error (&logging_error)`. (I realise the function returns immediately after this, but it would safeguard against double-setting the error if the function were to ever be extended.)

pwithnall commented on common/flatpak-transaction-log.c:
```diff
@@ -0,0 +1,305 @@
+/* flatpak-transaction-log.c
```
Nitpick: Do we have to include the filename in the header comment? It adds no value and is likely to be copy-pasted elsewhere and become incorrect.

smspillaz replied:
I'm not sure - it seems like this is a convention throughout flatpak generally? Full disclosure - I copypasted from `flatpak-db.c` here ;-)

pwithnall replied:
I guess @alexlarsson should answer this. Personally, I do not see any upsides to putting the filename at the top of the file.

pwithnall commented on common/flatpak-transaction-log.c:
```diff
@@ -0,0 +1,305 @@
+/* flatpak-transaction-log.c
+ *
+ * Copyright (C) 2017 Endless Mobile, Inc
```
Nitpick: Can we use a Unicode © pretty-please? :-)

pwithnall commented on common/flatpak-transaction-log.c:
```diff
@@ -0,0 +1,305 @@
+
+enum {
+  PROP_0,
+  PROP_PATH,
```
Nitpick: Drop `PROP_0` and just set `PROP_PATH = 1` to save defining an identifier which should never be used.

pwithnall commented on common/flatpak-transaction-log.c:
```diff
@@ -0,0 +1,305 @@
+{
+  FlatpakTransactionLog *self = (FlatpakTransactionLog *) object;
+
+  g_clear_pointer (&self->path, g_object_unref);
```
Use `g_clear_object (&self->path)` to be simpler.

pwithnall commented on common/flatpak-transaction-log.c:
```diff
@@ -0,0 +1,305 @@
+  object_class->get_property = flatpak_transaction_log_get_property;
+  object_class->set_property = flatpak_transaction_log_set_property;
+
+  g_object_class_install_property (object_class,
```
Add a gtk-doc comment for it (and for the rest of the API in the new class). I realise it’s not public API, but documentation is useful for whoever’s editing things inside flatpak.git too.

pwithnall commented on common/flatpak-transaction-log.h:
```diff
@@ -0,0 +1,57 @@
+
+#define FLATPAK_TYPE_TRANSACTION_LOG (flatpak_transaction_log_get_type ())
+#define FLATPAK_TRANSACTION_LOG(obj) (G_TYPE_CHECK_INSTANCE_CAST ((obj), FLATPAK_TYPE_TRANSACTION_LOG, FlatpakTransactionLog))
+#define FLATPAK_IS_TRANSACTION_LOG(obj) (G_TYPE_CHECK_INSTANCE_TYPE ((obj), FLATPAK_TYPE_TRANSACTION_LOG))
```
Simplify this all using `G_DECLARE_FINAL_TYPE`.

cosimoc commented on common/flatpak-transaction-log.c:
```diff
@@ -0,0 +1,305 @@
+                             GError       **error)
+{
+  g_autoptr(JsonParser) parser = json_parser_new ();
+  g_autoptr(GFileInputStream) stream = g_file_read (path, cancellable, error);
```
You could probably use `json_parser_load_from_file()` instead

smspillaz replied:
I had a look at that, though `json_parser_load_from_file` doesn't support cancellability. In any event, we'll need to hand-write a loader for this...

cosimoc commented on common/flatpak-transaction-log.c:
```diff
@@ -0,0 +1,305 @@
+    return NULL;
+
+  /* Attempting to parse an empty file does not result in an error, but instead
+   * results in a parser with no root. Handle this case accordingly */
```
Instead, you could simply return a new node from here instead of special-casing your own error later

smspillaz commented:
Okay, patchset updated and rebased. Main changes:
 - We no longer write strict RFC7519 JSON, instead we append to a log file with line-delimited JSON objects. Applications wishing to read the file should read it line-by-line and parse out each JSON object.
 - Added some gtkdoc
 - Fixed nitpicks

alexlarsson commented on common/flatpak-transaction-log.c:
```diff
@@ -0,0 +1,287 @@
+
+  json_builder_set_member_name (builder, "version");
+  json_builder_add_string_value (builder, EVENT_LOG_VERSION);
+  json_builder_end_object (builder);
```
I'm a bit uncertain about the version field. I mean, how would you use it? If we want to just add stuff to the json object that is fine without a version field. Consumers can just check for the new fields. A version is just useful if we're for some reason doing a completely backwards incompatible changes so that consumers of the existing versions have to ignore the new nodes. I think we should avoid such changes, and in the unlikely case we need them, just go with a different filename for the log.

smspillaz replied:
Okay - I guess if we only care about forward compatibility I can drop the field.

alexlarsson commented on common/flatpak-transaction-log.c:
```diff
@@ -0,0 +1,287 @@
+  JsonObject *object = json_node_get_object (base);
+
+  json_object_set_string_member (object, "ref", ref);
+  json_object_set_string_member (object, "origin", origin);
```
Can we also add the exact commit id we deployed?

pwithnall commented on common/flatpak-transaction-log.c:
```diff
@@ -0,0 +1,287 @@
+  return g_output_stream_write_all (G_OUTPUT_STREAM (stream),
+                                    (gconstpointer) serialized,
+                                    serialized_len,
+                                    &written,
```
`written` is never used.

smspillaz replied:
https://developer.gnome.org/gio/stable/GOutputStream.html#g-output-stream-write-all it does not look like the parameter is nullable?

pwithnall replied:
It is: https://git.gnome.org/browse/glib/tree/gio/goutputstream.c#n296

I’ll add the missing annotation in GLib.

pwithnall commented on common/flatpak-transaction-log.c:
```diff
@@ -0,0 +1,287 @@
+  json_builder_set_member_name (builder, "event");
+  json_builder_add_string_value (builder, event);
+
+  json_builder_set_member_name (builder, "version");
```
As Alex says, it’s pointless to add a `version` attribute before v2.

pwithnall commented on common/flatpak-transaction-log.c:
```diff
@@ -0,0 +1,287 @@
+deploy_install_event (const gchar *ref,
+                      const gchar *origin)
+{
+  JsonNode *base = build_base_event ("deploy-install");
```
This should be `g_autoptr(JsonNode)` and the `return` should use `g_steal_pointer()`. That makes the memory handling more obvious, and eliminates the possibility of bugs being introduced when refactoring.

pwithnall commented on common/flatpak-transaction-log.c:
```diff
@@ -0,0 +1,287 @@
+		                                                 const gchar           *ref,
+		                                                 const gchar           *origin,
+		                                                 GCancellable          *cancellable,
+		                                                 GError                **error)
```
Nitpick: looks like the argument alignment is off (unless it’s being messed up by github?)

smspillaz replied:
If you're referring to the alignment of `**error` - is the convention here that it is shifted back by one (such that the starting character of the identifier aligns?)

smspillaz replied:
Ah, looks like the tabs monster strikes again :(

pwithnall replied:
Yes, the starting characters of all the identifiers should align. And there should be no tabs; just spaces.

pwithnall replied:
👾

pwithnall commented on common/flatpak-transaction-log.c:
```diff
@@ -0,0 +1,287 @@
+ * flatpak_transaction_log_commit_uninstall_event:
+ * @log: a #FlatpakTransactionLog
+ * @ref: a decomposable Flatpak ref
+ * @origin: the origin of the flatpak to be installed, whether it is from
```
There is no `origin` argument to this function.

pwithnall commented on common/flatpak-transaction-log.c:
```diff
@@ -0,0 +1,287 @@
+ * @cancellable: a #GCancellable
+ * @error: a #GError
+ *
+ * Write an "install" event to the transaction log. The "ref" arguments will be
```
s/install/uninstall/

pwithnall commented on common/flatpak-transaction-log.c:
```diff
@@ -0,0 +1,287 @@
+   *
+   * The absolute path, on disk, where the transaction log will be written to.
+   * If the file indicated by the path, or any of its parent directories do
+   * not exist, they will be created.
```
Missing a `Since` line in the documentation comment. (`Since: 0.10.0`)

Same applies to the other new APIs here which should have gtk-doc comments.

pwithnall commented on common/flatpak-transaction-log.h:
```diff
@@ -0,0 +1,51 @@
+#include <gio/gio.h>
+#include <glib-object.h>
+
+#define FLATPAK_TYPE_TRANSACTION_LOG flatpak_transaction_log_get_type ()
```
Nitpick: Conventionally, this should be within the `G_BEGIN_DECLS` section. Stick it on the line above the `G_DECLARE_FINAL_TYPE`.

smspillaz replied:
Yeah, I wasn't exactly certain where it went in relation to `G_DECLARE_FINAL_TYPE`. I'll move it.

cosimoc commented on common/flatpak-dir.c:
```diff
@@ -5363,6 +5364,26 @@ flatpak_dir_deploy (FlatpakDir          *self,
+{
+  return g_file_get_child (dir->basedir, "transactions.log");
+}
+
```
Nitpick: extra empty line

cosimoc commented on common/flatpak-transaction-log.c:
```diff
@@ -0,0 +1,287 @@
+  /* Resize the buffer to accomadate one extra newline and write that
+   * in place at serialized_len */
+  serialized = json_generator_to_data (generator, &serialized_len);
+  serialized = g_realloc (serialized, (serialized_len + 1) * sizeof (gchar));
```
Instead of reallocating this, you could simply write the newline/terminator with a second call to `g_output_stream_write_all()`

smspillaz commented on common/flatpak-transaction-log.c:
```diff
@@ -0,0 +1,287 @@
+  stream = g_file_append_to (file,
+                             G_FILE_CREATE_NONE,
+                             cancellable,
+                             error);
```
Actually I think it should be sync in every case. If we made it async then there isn't any way to guarantee that events will be added to the transaction log in the same order that they occurred (plus, events might not be written if the program exits before the main loop has a chance to dispatch the write events).

pwithnall replied:
> If we made it async then there isn't any way to guarantee that events will be added to the transaction log in the same order that they occurred

That’s guaranteed by the ordering of dispatching idle events from the `GMainContext`. Might take a bit of care though.

> (plus, events might not be written if the program exits before the main loop has a chance to dispatch the write events).

That would be a bug in the program for not waiting for the `GMainContext` to finish handling any pending events before exiting.

---

I think the decision here is about where the I/O blocking happens: in a worker thread (in which case these functions should be sync); or in the main thread (in which case they should be async).

smspillaz replied:
I had a look at where these functions are being called, they are only ever called through the command line interface or through `flatpak_installation_install` / `flatpak_installation_uninstall`, both of which are sync functions already.

smspillaz commented on common/flatpak-transaction-log.c:
```diff
@@ -0,0 +1,287 @@
+  gsize serialized_len = 0;
+  gsize written = 0;
+  g_autoptr(JsonGenerator) generator = NULL;
+  g_autoptr(GError) my_error = NULL;
```
I've seen throughout flatpak_dir `my_error` is used, so I've just used that. But I can change to `local_error` if that was a mistake before.

pwithnall replied:
Fair enough, let’s keep using that convention within this file then.

smspillaz commented:
Rebased and updated. Main changes:

 - Fixed nitpicks
 - Removed `version` attribute
 - Changed `deploy-install` event to `deploy` - we now log all deploy events and don't attempt to distinguish between them (for now, at least).
 - Added `commit` attribute to `deploy` events

pwithnall commented on common/flatpak-transaction-log.c:
```diff
@@ -0,0 +1,293 @@
+ * newline delimited list of JSON-formatted object definitions which should
+ * be parsed one line at a time. This enables #FlatpakTransactionLog to write
+ * new entries in O(1) time, as opposed to having to read the entire log
+ * into memory so that it can be re-written again.
```
The JSON schema isn’t documented anywhere. Even if you don’t write an actual [JSON Schema](http://json-schema.org/) file at the moment, it would be useful to list the keys and types supported in a gtk-doc comment somewhere. (Maybe in the gtk-doc comment for the overall `FlatpakTransactionLog` class?)

pwithnall commented on common/flatpak-transaction-log.c:
```diff
@@ -0,0 +1,293 @@
+ * @log: a #FlatpakTransactionLog
+ * @ref: a decomposable Flatpak ref
+ * @origin: the origin of the flatpak to be installed, whether it is from
+ *          a bundle or from a remote.
```
Nitpick: `@commit` is missing from this gtk-doc comment.

pwithnall commented on common/flatpak-transaction-log.c:
```diff
@@ -0,0 +1,293 @@
+ * Returns: %TRUE if writing the event succeeded.
+ */
+gboolean
+flatpak_transaction_log_commit_deploy_event (FlatpakTransactionLog *log,
```
Nitpick: By convention, method implementations for GObject classes always call their first parameter `self`.

alexlarsson commented:
@rh-atomic-bot r+ 498a373

alexlarsson commented:
Seems to be a double free somewhere...

ramcq commented:
I'm late to the party but if we're revising this... can we also log the time of the event, the user who originated it, and the program name (ie g_get_prgname) just for a bit of context? :)

ramcq commented:
(Also given that remotes can be removed after an event is logged, we should probably record the collection ID if it exists?)

smspillaz commented:
@alexlarsson Just a heads up that we may not actually end up using this approach for the feature that we want to implement in our system updater (I misunderstood one of the requirements and it turns out that implementing it using a transaction log alone would be tricky. @ramcq had a better idea but it would be implemented within the context of our updater itself).

If you think this change would still be beneficial for the project at large, I'd be happy to finish it off, otherwise I wouldn't want to push a maintenance burden on to flatpak for a feature that wasn't being used. Thoughts?

alexlarsson commented:
I'd rather wait until we have a real user then. Lets close this for now. It'll be here for repopen if we need it.
